packages/flutter/lib/src/gestures/mouse_cursor.dart

```diff
@@ -0,0 +1,332 @@
+
+  @override
+  int get hashCode => hashValues(device, systemShape);
+
+  @override
+  String toString() {
+    return '$runtimeType(device: $device, systemShape: $systemShape)';
+  }
+}
+
+/// An interface for the operations that a [MouseCursor] can use to control the
+/// platform.
+///
+/// This interface is implemented by each platform that supports mouse cursor.
+///
+/// See also:
+///
+///  * [MouseCursor], whose subclasses and methods use this class to perform
+///    operations.
+///  * [MouseCursorManager], which takes this class as a parameter.
+abstract class MouseCursorPlatformDelegate {
+  /// Create a [MouseCursorPlatformDelegate].
+  const MouseCursorPlatformDelegate();
+
+  /// Asks the platform to change the cursor of `device` to the system cursor
+  /// specified by `systemShape`.
+  ///
+  /// It resolves to `true` if the operation is successful, `false` if the
+  /// operation is unsupported by the platform, or rejects to error if the
+  /// operation is implemented but an error occurs.
+  Future<bool> activateSystemCursor(MouseCursorPlatformActivateSystemCursorDetails details);
+}
+
+/// Details for [MouseCursor.activate], such as the target device and the
+/// platform.
+@immutable
+class MouseCursorActivateDetails {
+  /// Create details for a [MouseCursor.activate] call.
+  ///
+  /// All parameters must not be null.
+  const MouseCursorActivateDetails({
+    @required this.device,
+    @required this.platformDelegate,
+  }) : assert(device != null), assert(platformDelegate != null);
+
+  /// The pointer device that should change cursor.
+  final int device;
+
+  /// The delegate of the platform that the program is currently running on,
+  /// with which the cursor can perform operations related to mouse cursor.
+  final MouseCursorPlatformDelegate platformDelegate;
+}
+
+/// An interface for mouse cursor definitions.
+///
+/// A mouse cursor is a graphical image on the screen that echoes the movement
+/// of a pointer device, such as a mouse or a stylus. A [MouseCursor] object
+/// is a stateless definition of a kind of mouse cursor, such as an arrow,
+/// a pointing hand, or an I-beam.
+///
+/// A [MouseCursor] object is used by being assigned to a region. The most
+/// common way of doing so is [MouseRegion], which is wrapped and exposed by
+/// many other widgets.
+///
+/// When a pointer enters a region that is assigned with a [MouseCursor], the
+/// cursor's [MouseCursor.activate] is called.
+///
+/// See also:
+///
+///  * [MouseRegion], which is a common way of assigning a region with a
+///    [MouseCursor].
+///  * [MouseTracker], which determines the cursor that each device should show,
+///    and dispatches the changing callbacks.
+///  * [SystemMouseCursors], which provies many system cursors.
+///  * [NoopMouseCursors], which is a special type of mouse cursor that does
+///    nothing.
+@immutable
+abstract class MouseCursor {
+  /// Create a mouse cursor.
+  ///
+  /// The base constructor does nothing.
```

Comment:
for no-argument `const` constructors, please use the boilerplate that we use for all no-argument `const` constructors.

(this applies to other classes in this PR as well)